File: frontend/test/core/contracts/enums_generated_test.dart
// M0 contract freeze — round-trip enum canonical + UUIDv7 contract.
// Nguồn: shared/contracts/enums.json · Xem M0-contract-freeze.md §Test plan.
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:frontend/core/contracts/enums.generated.dart';

Map<String, dynamic> _readJson(String relFromRepoRoot) {
  // `flutter test` chạy với cwd = frontend/ ; repo root là thư mục cha.
  final f = File('../$relFromRepoRoot');
  return jsonDecode(f.readAsStringSync()) as Map<String, dynamic>;
}

void main() {
  final src = _readJson('shared/contracts/enums.json');
  final idFixtures = _readJson('shared/contracts/fixtures/id-samples.json');
  final srcEnums = src['enums'] as Map<String, dynamic>;

  List<String> values(String name) =>
      (srcEnums[name]['values'] as List).cast<String>();

  final cases = <String, ({List<String> wires, String Function(String) parse})>{
    'workspace_lifecycle_stage': (
      wires: WorkspaceLifecycleStage.values.map((e) => e.wire).toList(),
      parse: (v) => WorkspaceLifecycleStage.fromWire(v).toApi(),
    ),
    'project_lifecycle_stage': (
      wires: ProjectLifecycleStage.values.map((e) => e.wire).toList(),
      parse: (v) => ProjectLifecycleStage.fromWire(v).toApi(),
    ),
    'workspace_status': (
      wires: WorkspaceStatus.values.map((e) => e.wire).toList(),
      parse: (v) => WorkspaceStatus.fromWire(v).toApi(),
    ),
    'project_status': (
      wires: ProjectStatus.values.map((e) => e.wire).toList(),
      parse: (v) => ProjectStatus.fromWire(v).toApi(),
    ),
    'runtime_mode': (
      wires: RuntimeMode.values.map((e) => e.wire).toList(),
      parse: (v) => RuntimeMode.fromWire(v).toApi(),
    ),
    'sync_policy': (
      wires: SyncPolicy.values.map((e) => e.wire).toList(),
      parse: (v) => SyncPolicy.fromWire(v).toApi(),
    ),
    'sync_status': (
      wires: SyncStatus.values.map((e) => e.wire).toList(),
      parse: (v) => SyncStatus.fromWire(v).toApi(),
    ),
    'legal_entity_status': (
      wires: LegalEntityStatus.values.map((e) => e.wire).toList(),
      parse: (v) => LegalEntityStatus.fromWire(v).toApi(),
    ),
  };

  group('workspace-canonical enums (generated dart)', () {
    cases.forEach((name, c) {
      test('$name: khớp thứ tự value nguồn', () {
        expect(c.wires, equals(values(name)));
      });
      test('$name: round-trip mọi value', () {
        for (final v in values(name)) {
          expect(c.parse(v), equals(v));
        }
      });
      test('$name: value lạ -> throw, không map ngầm', () {
        expect(() => c.parse('__NOT_REAL__'), throwsArgumentError);
      });
    });

    test('tryFromWire trả null cho value lạ (không default ngầm)', () {
      expect(WorkspaceLifecycleStage.tryFromWire('S0_GENESIS'), isNull);
      expect(ProjectLifecycleStage.tryFromWire('S0_EXPLORE'), isNull);
    });

    test('stage enum không lẫn mã legacy S*', () {
      expect(
        WorkspaceLifecycleStage.values.every((e) => e.wire.startsWith('W')),
        isTrue,
      );
      expect(
        ProjectLifecycleStage.values.every((e) => e.wire.startsWith('P')),
        isTrue,
      );
    });

    test('migration map phủ đủ target canonical', () {
      expect(
        legacyWorkspaceStageToCanonical.values.toSet(),
        equals(WorkspaceLifecycleStage.values.map((e) => e.wire).toSet()),
      );
      expect(
        legacyProjectStageToCanonical.values.toSet(),
        equals(ProjectLifecycleStage.values.map((e) => e.wire).toSet()),
      );
    });
  });

  group('ID serialization contract (M0)', () {
    test('Snowflake decimal string round-trip qua BigInt, không mất precision', () {
      final samples = (idFixtures['snowflake_decimal_strings']['samples'] as List)
          .cast<String>();
      for (final s in samples) {
        final decoded =
            jsonDecode(jsonEncode({'workspace_id': s})) as Map<String, dynamic>;
        expect(decoded['workspace_id'], isA<String>());
        expect(BigInt.parse(decoded['workspace_id'] as String).toString(), equals(s));
      }
    });

    test('giá trị 63-bit vỡ nếu ép qua double', () {
      final risky = (idFixtures['snowflake_decimal_strings']
              ['must_not_equal_after_double_roundtrip'] as List)
          .cast<String>();
      for (final s in risky) {
        final asDouble = double.parse(s);
        expect(asDouble.toStringAsFixed(0) == s, isFalse);
        expect(BigInt.parse(s).toString(), equals(s));
      }
    });

    test('UUIDv7 fixtures: version 7 + variant 10xx + canonical', () {
      final re = RegExp(
        r'^[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$',
      );
      for (final s in (idFixtures['uuidv7']['ordered_samples'] as List).cast<String>()) {
        expect(re.hasMatch(s), isTrue, reason: s);
      }
      for (final s in (idFixtures['uuidv7']['not_v7'] as List).cast<String>()) {
        expect(re.hasMatch(s), isFalse, reason: s);
      }
    });

    test('UUIDv7 đơn điệu thời gian = sắp xếp lexicographic', () {
      final ordered =
          (idFixtures['uuidv7']['ordered_samples'] as List).cast<String>();
      final sorted = [...ordered]..sort();
      expect(sorted, equals(ordered));
    });
  });
}

File: services/cosa/shared/contracts/enums.generated.test.ts
// M0 contract freeze — round-trip cho enum canonical + ID serialization contract.
// Nguồn: shared/contracts/enums.json · Xem M0-contract-freeze.md §Test plan.
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { describe, expect, it } from "vitest";
import * as enums from "./enums.generated";

const REPO_ROOT = resolve(__dirname, "../../../..");
const src = JSON.parse(
  readFileSync(resolve(REPO_ROOT, "shared/contracts/enums.json"), "utf8"),
);
const idFixtures = JSON.parse(
  readFileSync(resolve(REPO_ROOT, "shared/contracts/fixtures/id-samples.json"), "utf8"),
);

const CASES: Record<string, { arr: readonly string[]; parse: (v: string) => string }> = {
  workspace_lifecycle_stage: {
    arr: enums.WORKSPACE_LIFECYCLE_STAGE,
    parse: enums.parseWorkspaceLifecycleStage,
  },
  project_lifecycle_stage: {
    arr: enums.PROJECT_LIFECYCLE_STAGE,
    parse: enums.parseProjectLifecycleStage,
  },
  workspace_status: { arr: enums.WORKSPACE_STATUS, parse: enums.parseWorkspaceStatus },
  project_status: { arr: enums.PROJECT_STATUS, parse: enums.parseProjectStatus },
  runtime_mode: { arr: enums.RUNTIME_MODE, parse: enums.parseRuntimeMode },
  sync_policy: { arr: enums.SYNC_POLICY, parse: enums.parseSyncPolicy },
  sync_status: { arr: enums.SYNC_STATUS, parse: enums.parseSyncStatus },
  legal_entity_status: {
    arr: enums.LEGAL_ENTITY_STATUS,
    parse: enums.parseLegalEntityStatus,
  },
};

describe("workspace-canonical enums (generated)", () => {
  for (const [name, c] of Object.entries(CASES)) {
    it(`${name}: khớp thứ tự value nguồn`, () => {
      expect([...c.arr]).toEqual(src.enums[name].values);
    });
    it(`${name}: round-trip mọi value`, () => {
      for (const v of src.enums[name].values) expect(c.parse(v)).toBe(v);
    });
    it(`${name}: value lạ -> throw, không map ngầm về default`, () => {
      expect(() => c.parse("__NOT_REAL__")).toThrow(/Unknown/);
    });
  }

  it("stage enum không lẫn mã legacy S*", () => {
    expect(enums.WORKSPACE_LIFECYCLE_STAGE.every((v) => v.startsWith("W"))).toBe(true);
    expect(enums.PROJECT_LIFECYCLE_STAGE.every((v) => v.startsWith("P"))).toBe(true);
  });

  it("migration map phủ đủ target canonical", () => {
    expect(new Set(Object.values(enums.LEGACY_WORKSPACE_STAGE_TO_CANONICAL))).toEqual(
      new Set(enums.WORKSPACE_LIFECYCLE_STAGE),
    );
    expect(new Set(Object.values(enums.LEGACY_PROJECT_STAGE_TO_CANONICAL))).toEqual(
      new Set(enums.PROJECT_LIFECYCLE_STAGE),
    );
  });
});

describe("ID serialization contract (M0)", () => {
  it("Snowflake luôn là decimal string; giá trị 63-bit vỡ nếu dùng Number", () => {
    for (const s of idFixtures.snowflake_decimal_strings.samples) {
      expect(BigInt(s).toString()).toBe(s);
    }
    for (const s of idFixtures.snowflake_decimal_strings
      .must_not_equal_after_double_roundtrip) {
      expect(String(Number(s))).not.toBe(s); // đây là lý do cấm serialize dạng number
      expect(BigInt(s).toString()).toBe(s); // string thì an toàn
    }
  });

  it("JSON.parse giữ Snowflake ở dạng string", () => {
    const decoded = JSON.parse(JSON.stringify({ workspace_id: "9223372036854775807" }));
    expect(typeof decoded.workspace_id).toBe("string");
  });

  it("UUIDv7 fixtures: đúng version 7, variant 10xx, canonical", () => {
    const re =
      /^[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/;
    for (const s of idFixtures.uuidv7.ordered_samples) expect(s).toMatch(re);
    for (const s of idFixtures.uuidv7.not_v7) expect(s).not.toMatch(re);
  });

  it("UUIDv7 đơn điệu thời gian = sắp xếp lexicographic", () => {
    const o = idFixtures.uuidv7.ordered_samples as string[];
    expect([...o].sort()).toEqual(o);
  });
});
